### PR TITLE
Add test page

### PR DESCRIPTION
**Pull request by SeoJungHong**
- 그림판 구현 : mnist/test/draw 에 그림판 기능 구현함. 280 x 280 크기의 캔버스에 그림을 그리고 'set'을 누르면, 이미지 정보를 28 x 28픽셀에 맞게 줄여서 흑백 데이터에 맞게 가공한 뒤 `draw_result` api를 호출한다. draw_result에서는  mnist_data형식에 맞게 포매팅 해서(utils.py 참고) services.py 의 `test_single`을 호출한다. 결과값은 아직은 시각적으로 가공하지 않고 일단은 결과값을 화면에 0~9까지 출력하고 있다. 이 때 기존에 트레이닝 한 모델을 로드해서 사용하는데, 이는 유저가 최근에 쿠키에 세팅한 값을 기준으로 결정된다. 추후 현재 모델 정보를 같이 출력하거나, 선택할 수 있게 수정해야 할것 같다. 자세한 것은 아래의 `모델 저장 및 로드` 참조.

- `services.py` 에서 백엔드 로직 수정사항
    - static variable(class끼리 고정적임. 파이썬 문법에서 아무것도 안붙은 멤버 변수는 static), class variable(인스턴스마다 다름. self.{variable} 의 형식으로 쓰임) 구분
    - tensorboard 표시를 위해 name(노드 내부 변수명), name_scope(그래프의 노드) 수정
    - 유저 선택에 따라 weight_initialize(xavier), multi layer model 변경 지원 -> view에서 전달해주는 쿠키 정보를 토대로 변경되게 해 둠. (`get_model` 메소드) 기존 num_of_layers 를 일단은 single vs multiple 로 변경하였다.
    - 파라미터를 명시적으로 받아서 이용하도록 변경. ex. param[0] -> learning_rate

- 모델 저장 및 로드 : test_single 을 구현하기 위해, tensorflow에서 제공하는 Saver()를 써서 현재 모델의 Variable들의 상태를 저장하고, 로드하는 기능을 이용하였다. (http://dsmoon.tistory.com/entry/TensorFlow-Variables 참고) 일단은 모델을 log처럼 임시로 로컬에 저장하도록 해 두었다. 모델을 로드하기 전에 어떤 타입의 모델을 로드해야 하는지 결정해서 변수 타입을 Saver에 전달해야 하기 때문에, 현재 쿠키 세팅 값 중 model_type(기존 레이어 수), weight_initialize 여부 이 두가지의 변수를 통해 모델의 타입을 결정해서 이를 저장 파일 이름에 반영하고, 로드하도록 하였다. 이 기능이 set_algorithm, test_all, test_single 모두에 반복적으로 쓰이기 대문에 get_model() 이라는 메소드 하나로 합쳐두었다.

- 테스트 케이스 추가(`tests.py`)
    - 단일 레이어 트레이닝(`testBasicTraining`), 다중 레이어 트레이닝(`testAdvancedTraining`), 전체 테스트(즉, 테스트 데이터 이용 - `testTestAll`), 단일 테스트(그림판을 위해 구현. 특정 이미지 데이터에 대해 분석값 리턴 - `testTestSingle`) 에 관한 테스트 케이스를 만들었다.

변경 사항이 좀 많아서...위의 내용 참고하시고 제가 추가한 테스트 케이스를 돌려보시면 좋을 것 같습니다. 결국 저희가 구현해야 할 큰 흐름에 맞추어(Training, test_all, test_single) 기본적으로 메인 로직이 어떻게 돌아가도록 제가 의도했는지 파악하기 쉬울 것 같습니다.

### CODE REVIEW

File: practice/templates/practice/base.html [hunk 1]
```diff
@@ -22,21 +22,21 @@
     <div class=" w3-sidenav w3-center" id="menu">
       <ul class="step">STEP</ul>
       <ul class="mainmenu">
-        <li><i class="material-icons icon">folder_open</i><p>Data</p></a></li>
+        <li><i class="material-icons icon">folder_open</i><p>Data</p><</li>
```
Comment (hyunMin11) [inline practice/templates/practice/base.html]: 아이쿠 제가 </a>를 다 안뺐었네여. 여기 <<로 오타가 있어요.
제가 할 부분이었는데 죄송합니다. 그리고 algorithm부분 구현하신거 정말 수고하셨어요 확인 했는데 구성이 잘 된것 같습니다.

File: practice/templates/practice/test/draw.html [hunk 2]
```diff
@@ -0,0 +1,121 @@
...
+    <p>Draw a number</p>
+    <p>
+        <canvas id="input" width="280" height="280" style="border:solid 2px"></canvas>
+    </p>
+    <p>
+        <button id="clear" class="btn btn-default">clear</button>
+        <button id="set" class="btn btn-default">set</button>
+    </p>
+</div>
+
```
Comment (hyunMin11) [inline practice/templates/practice/test/draw.html]: 여기는 굳이 띄지 않아도 될 것 같습니다.

File: practice/templates/practice/test/draw.html [hunk 3]
```diff
@@ -0,0 +1,121 @@
+{% extends 'practice/base_test.html' %}
+{% load staticfiles %}
+
+{% block frame %}
+
```
Comment (hyunMin11) [inline practice/templates/practice/test/draw.html]: 제가 다른 file에서 이 부분은 다 붙여놔서 혹시 붙여주실 수 있으신가요?

File: practice/templates/practice/test/draw.html [hunk 4]
```diff
@@ -0,0 +1,121 @@
...
+        <canvas id="input" width="280" height="280" style="border:solid 2px"></canvas>
+    </p>
+    <p>
+        <button id="clear" class="btn btn-default">clear</button>
+        <button id="set" class="btn btn-default">set</button>
+    </p>
+</div>
+
+<div class="col-md-7">
+    <table id="result" class="table">
```
Comment (hyunMin11) [inline practice/templates/practice/test/draw.html]: 2space로 implement해주세요

File: practice/templates/practice/test/draw.html [hunk 5]
```diff
@@ -0,0 +1,121 @@
...
+                    console.log('result: ' + data.results);
+                    for (var i = 0; i < 10; i++) {
+                        $('#result').find('tr').eq(i).find('td').text(data.results[i]);
+                    }
+                }
+            );
+        };
+        img.src = canvas.toDataURL();
+    }
+
```
Comment (hyunMin11) [inline practice/templates/practice/test/draw.html]: 이곳도 붙여주시는것이 좋을것 같아요^^

File: practice/utils.py [hunk 6]
```diff
@@ -0,0 +1,8 @@
+import numpy as np
+
+
```
Comment (hyunMin11) [inline practice/utils.py]: 한 줄 더 띄여졌습니다.

File: practice/utils.py [hunk 7]
```diff
@@ -0,0 +1,8 @@
...
+def image_to_mnist(image_data):
+    return ((255 - np.array(image_data, dtype=np.int)) / 255.0).reshape(1, 784)
```
Comment (hyunMin11) [inline practice/utils.py]: reshape에서 784를 IMAGE_PIXELS와 같은 상수로 선언하면 보기 편할것 같아요!!

File: practice/templates/practice/base.html (hunk 1 above)
Comment (SeoJungHong) [inline practice/templates/practice/base.html]: @hyunMin11 으앗ㅋㅋㅋ '<' 얘가 남아있었군요 빨간 줄 안그어져서 못봤는데... Data, Algorithm 메뉴에 < 이게 떠있길래 현재 있는 위치를 표시해주는 UI 효과인줄 알았습니닼ㅋㅋㅋ바꿀게요!

File: practice/tests.py [hunk 8]
```diff
@@ -5,29 +5,28 @@
...
...255,255,255,255,255,255,255,227,126,255,255,255,255,255,255,255,255,255,255,255,255,255,255,255,255,255,255,255,255,255,255,255,255,255,255,0,255,255,255,255,255,255,255,255,255,255,255,255,255,255,255,255,255,255,255,255,255,255,255,255,255,255,255,0,255,255,255,255,255,255,255,255,255,255,255,255,255,255,255,255,255,255,255,255,255,255,255,255,255,255,228,204,255,255,255,255,255,255,255,255,255,255,255,255,255,255,255,255,255,255,255,255,255,255,255,255,255,255,255,255,255,255,255,255,255,255,255,255,255,255,255,255,255,255,255,255,255,255,255,255,255,255,255,255,255,255,255,255,255,255,255,255,255,255,255,255,255,255,255,255,255,255,255,255,255,255,255,255,255,255,255,255,255,255,255,255,255,255,255,255,255,255,255,255,255,255,255,255,255,255,255,255,255,255,255,255,255,255,255,255,255,255,255,255,255,255,255,255,255,255,255,255,255,255,255,255,255,255,255,255,255,255,255,255,255,255,255,255,255,255,255,255,255,255,255,255,255,255,255,255,255,255,255,255,255,255,255,255,255,255,255,255,255,255,255,255,255,255,255,255,255,255,255,255,255,255,255,255,255,255,255,255,255,255,255,255,255,255,255,255,255,255,255,255,255,255,255,255,255,255,255,255,255,255,255,255,255,255,255,255,255]
```
Comment (hyunMin11) [inline practice/tests.py]: 255가 여러번 나오는것 같은데 혹시 상수로 선언하는 게 어떨까요??

File: practice/views.py [hunk 9]
```diff
@@ -39,18 +27,34 @@ class Algorithm(APIView):
     template_name = 'practice/algorithm.html'
 
     def get(self, request, practice_name):
-        setting_list = MNIST.get_algorithm_settings()
+        setting_list = {
+            'Model Type': [
+                'Single', 'Multiple'
```
Comment (hyunMin11) [inline practice/views.py]: Model Type이름을 Single layer, Multiple layers로 해서 더 직관적으로 나타내게 하면 좋을것 같은데 어떻게 생각하시나요??

File: practice/services.py [hunk 10]
```diff
@@ -42,131 +36,150 @@ def tensorboard():
...
 
+    # Static Variables
     LOGS_PATH = 'practice/.logs'
     DATA_PATH = 'practice/.data'
-
-    # Data
-    training_data = None
-    test_data = None
+    MODEL_PATH = 'practice/.models/mnist'
+    BATCH_SIZE = 100
```
Comment (hyunMin11) [inline practice/services.py]: IMAGE_SIZE = 28
IMAGE_PIXELS = IMAGE_SIZE * IMAGE_SIZE
NUM_CLASSES = 10
ONE_BYTE = 255
이런식으로 상수 선언해서 사용하면 좋을 듯 싶습니다. 제가 이전 merge시 변경했었는데 누락된것 같군요

File: practice/templates/practice/test/draw.html (hunk 3 above)
Comment (SeoJungHong) [inline practice/templates/practice/test/draw.html]: 넹넹!

File: practice/tests.py (hunk 8 above)
Comment (SeoJungHong) [inline practice/tests.py]: @hyunMin11 아 저 데이터는 제가 그림판에 임의로 7이라고 쓴 숫자에 대한 하나의 예시에요. 그냥 수많은 올바른 인풋 중 하나의 값으로 적어둔거라서 상수로 선언할 필요까지는 없을 것 같아요. 그림판에 한번 그려보시면 저런 비슷한 데이터를 얻을 수 있어요

File: practice/utils.py (hunk 7 above)
Comment (SeoJungHong) [inline practice/utils.py]: 좋은 생각이네요 앞서 MNIST class쪽도 같이 바꿀게요

File: practice/views.py (hunk 9 above)
Comment (SeoJungHong) [inline practice/views.py]: 좋습니다. 반영할게요~

**Comment by hyunMin11** [comment]
아직 Activation Function은 적용이 안된건 같네요!! 정말 수고하셨어요!! style같은 작은 부분만 정리하면 좋을 것 같습니다~

File: practice/templates/practice/test/draw.html (hunk 4 above)
Comment (SeoJungHong) [inline practice/templates/practice/test/draw.html]: 네 반영햇습니다~

File: practice/utils.py (hunk 7 above)
Comment (SeoJungHong) [inline practice/utils.py]: import loop error가 나서.. 일단 여기는 그대로 두겠습니다. 다른 부분은 상수 참조하도록 했어요!

**Comment by SeoJungHong** [comment]
@hyunMin11 넹! 알고리즘 코드 반영하는것 관련해서 남은 과제는 저희 태스크 관리하는 엑셀파일에 적어놨어요. activation function, dropout, number of layers 는 아직이에요

**Comment by hyunMin11** [comment]
확인했습니다 머지해주세요~
